File: backend/core/dependency_specs.py
"""Parse registry ``distribution.dependencies`` entries."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class DependencySpec:
    model_id: str
    version: str | None = None


def parse_dependencies(raw: Any) -> list[DependencySpec]:
    if not isinstance(raw, list):
        return []
    out: list[DependencySpec] = []
    for item in raw:
        if isinstance(item, str):
            model_id = item.strip()
            if model_id:
                out.append(DependencySpec(model_id=model_id))
            continue
        if isinstance(item, dict):
            model_id = str(item.get("model_id") or item.get("model") or "").strip()
            if not model_id:
                continue
            version = item.get("version")
            version_key = str(version).strip() if isinstance(version, str) and version.strip() else None
            out.append(DependencySpec(model_id=model_id, version=version_key))
    return out


def dependency_model_ids(raw: Any) -> list[str]:
    return [spec.model_id for spec in parse_dependencies(raw)]

File: frontend/src/utils/dependencySpecs.ts
/** Registry ``distribution.dependencies`` — string model id or versioned object. */

export interface DependencySpec {
  modelId: string;
  version: string | null;
}

export type RegistryDependency =
  | string
  | {
      model_id?: string;
      model?: string;
      version?: string;
    };

export function parseDependencies(raw: unknown): DependencySpec[] {
  if (!Array.isArray(raw)) return [];
  const out: DependencySpec[] = [];
  for (const item of raw) {
    if (typeof item === 'string') {
      const modelId = item.trim();
      if (modelId) out.push({ modelId, version: null });
      continue;
    }
    if (item && typeof item === 'object') {
      const obj = item as Record<string, unknown>;
      const modelId = String(obj.model_id ?? obj.model ?? '').trim();
      if (!modelId) continue;
      const versionRaw = obj.version;
      const version =
        typeof versionRaw === 'string' && versionRaw.trim() ? versionRaw.trim() : null;
      out.push({ modelId, version });
    }
  }
  return out;
}

export function isDependencyReady(
  spec: DependencySpec,
  detailedStatus: Record<string, { versions?: Record<string, { ready?: boolean }> }> | undefined,
  modelReady: (modelId: string) => boolean,
): boolean {
  if (!spec.version) {
    return modelReady(spec.modelId);
  }
  const versions = detailedStatus?.[spec.modelId]?.versions;
  if (versions?.[spec.version]?.ready) return true;
  // ``shared`` encoders satisfied when full ``original`` is installed.
  if (spec.version === 'shared' && versions?.original?.ready) return true;
  return false;
}
